src/natcap/invest/scenic_quality/scenic_quality.py

```diff
@@ -458,13 +406,186 @@ def execute(args):
     graph.join()
 
 
+def _determine_valid_viewpoints(dem_path, structures_path):
+    """Determine which viewpoints are valid and return them.
+
+    A point is considered valid when it meets all of these conditions:
+
+        1. The point must be within the bounding box of the DEM
+        2. The point must not overlap a DEM pixel that is nodata
+        3. The point must not have the same coordinates as another point
+
+    All invalid points are skipped, and a logger message is written for the
+    feature.
+
+    Args:
+        dem_path (str): The path to a GDAL-compatible digital elevation model
+            raster on disk.  The projection must match the projection of the
+            ``structures_path`` vector.
+        structures_path (str): The path to a GDAL-compatible vector containing
+            point geometries and, optionally, a few fields describing
+            parameters to the viewshed:
+
+                * 'RADIUS' or 'RADIUS2': How far out from the viewpoint (in m)
+                    the viewshed operation is permitted to extend. Default: no
+                    limit.
+                * 'HEIGHT': The height of the structure (in m). Default: 0.0
+                * 'WEIGHT': The numeric weight that this viewshed should be
+                    assigned when calculating visual quality. Default: 1.0
+
+    Returns:
+        An unsorted list of the valid viewpoints and their metadata.  The
+        tuples themselves are in the order::
+
+            (viewpoint, radius, weight, height)
+
+        Where
+
+            * ``viewpoint``: a tuple of
+                ``(projected x coord, projected y coord``)
+            * ``radius``: the maximum radius of the viewshed
+            * ``weight``: the weight of the viewshed (for calculating visual
+                quality)
+            * ``height``: The height of the structure at this point.
+    """
+    dem_raster_info = pygeoprocessing.get_raster_info(dem_path)
+    dem_nodata = dem_raster_info['nodata'][0]
+    dem_gt = dem_raster_info['geotransform']
+    bbox_minx, bbox_miny, bbox_maxx, bbox_maxy = (
+        dem_raster_info['bounding_box'])
+
+    # Use interleaved coordinates (xmin, ymin, xmax, ymax)
+    spatial_index = rtree.index.Index(interleaved=True)
+
+    structures_vector = gdal.OpenEx(structures_path, gdal.OF_VECTOR)
+    for structures_layer_index in range(structures_vector.GetLayerCount()):
+        structures_layer = structures_vector.GetLayer(structures_layer_index)
+        layer_name = structures_layer.GetName()
+        LOGGER.info('Layer %s has %s features', layer_name,
+                    structures_layer.GetFeatureCount())
+
+        fieldnames = set(
+            column.GetName() for column in structures_layer.schema)
+        radius_fieldname = None
+        for possible_radius_fieldname in ('RADIUS', 'RADIUS2'):
+            if possible_radius_fieldname in fieldnames:
+                radius_fieldname = possible_radius_fieldname
+                break
```

Comment:
I for some reason thought it'd be fun and maybe more clean to use `set` operations here:
```
possible_radius_fieldname  = set(['RADIUS', 'RADIUS2']).intersection(fieldnames)
if possible_radius_fieldnames:
    radius_fieldname = possible_radius_fieldnames.pop()
```
Feel free to ignore.

Reply:
Hey, that's a neat solution!  Added in 8e29e443